cluster-autoscaler/core/scaleup/resource_manager.go

```diff
@@ -0,0 +1,323 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scaleup
+
+import (
+	"fmt"
+	"math"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/core/utils"
+	ca_processors "k8s.io/autoscaler/cluster-autoscaler/processors"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/klog/v2"
+	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+// ScaleUpLimitUnknown is used as a value in ScaleUpResourcesLimits if actual limit could not be obtained due to errors talking to cloud provider.
+const ScaleUpLimitUnknown = math.MaxInt64
+
+// ScaleUpResourceManager provides resource checks before scaling up the cluster.
+type ScaleUpResourceManager interface {
+	ComputeScaleUpResourcesDelta(nodeInfo *schedulerframework.NodeInfo, nodeGroup cloudprovider.NodeGroup) (ScaleUpResourcesDelta, errors.AutoscalerError)
+	CheckScaleUpDeltaWithinLimits(delta ScaleUpResourcesDelta) ScaleUpLimitsCheckResult
+	ApplyScaleUpResourcesLimits(newNodes int, nodeGroup cloudprovider.NodeGroup, nodeInfo *schedulerframework.NodeInfo) (int, errors.AutoscalerError)
+}
+
+type scaleUpResourceManager struct {
+	context    *context.AutoscalingContext
+	processors *ca_processors.AutoscalingProcessors
+	nodes      []*apiv1.Node
+	nodeInfos  map[string]*schedulerframework.NodeInfo
+	nodeGroups []cloudprovider.NodeGroup
+
+	nodesFromNotAutoscaledGroups []*apiv1.Node
+	resourceLimiter              *cloudprovider.ResourceLimiter
+	scaleUpResourcesLeft         ScaleUpResourcesLimits
+}
+
+// ScaleUpLimitsCheckResult contains the limit check result and the exceeded resources if any.
+type ScaleUpLimitsCheckResult struct {
+	Exceeded          bool
+	ExceededResources []string
+}
+
+// ScaleUpResourcesLimits is a map: the key is resource type and the value is resource limit.
+type ScaleUpResourcesLimits map[string]int64
+
+// ScaleUpResourcesDelta is a map: the key is resource type and the value is resource delta.
+type ScaleUpResourcesDelta map[string]int64
+
+// NewScaleUpResourceManager creates an instance of scale up resource manager with provided parameters.
+func NewScaleUpResourceManager(context *context.AutoscalingContext, processors *ca_processors.AutoscalingProcessors, nodes []*apiv1.Node,
```

Comment:
I would consider keeping only `processors` as member in the struct, so that it can be created once and reused between scale ups. This would be similar to how resource limits are handled in scale down: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/scaledown/resource/limits.go